### PR TITLE
dynamic schema selection based on dict keys

### PR DESCRIPTION
**Pull request by radix**
Often times when `anyof` (or similar rules) are used, what we really want to do is *select* a schema based on dict keys.

This PR introduces two different options for this.

### when_key_is

Use this when you have dictionaries that have a fixed key, such as `"type"`, which specifies some specific format to use. For example, if you have data that can look like this:

```json
{"type": "elephant", "trunk_length": 60}
{"type": "eagle", "wingspan": 50}
```

Then you would use `when_key_is`, like this:

```json
{
    "type": "dict",
    "when_key_is": {
        "key": "type",
        "choices": {
            "elephant": {
                "schema": {"trunk_length": {"type": "integer"}}
            },
            "eagle": {
                "schema": {"wingspan": {"type": "integer"}}
            },
        }
    }
}
```

### when_key_exists

Use this when you have dictionaries where you must choose the schema based on keys that exist in the data exclusively for their type of data. For example, if you have data that can look like this:

```json
{"image_url": "foo.jpg", "width": 30}
{"color": "red"}
```

Then you would use `when_key_exists`, like this:

```json
{
    "type": "dict",
    "when_key_exists": {
        "image_url": {
            "schema": {"image_url": {"type": "string"}, "width": {"type": "integer"}}
        },
        "color": {
            "schema": {"color": {"type": "string"}}
        },
    }
}
```

## Questions

- Does anyone have suggestions for better names than `when_key_is` and `when_key_exists`?
- Or better names for `DictWhenKeyIs` and `DictWhenKeyExists`?

### CODE REVIEW

File: sureberus/schema.py [hunk 1]
```diff
@@ -11,11 +11,19 @@ def mk(d, kw, **morekw):
...
+        schema = {}
     if anyof is not None:
         kwargs['anyof'] = anyof
     return mk(None, kwargs, type='dict', schema=schema, required=required)
 
+def DictWhenKeyIs(key, choices, **kwargs):
+    return Dict(when_key_is={'key': key, 'choices': choices}, **kwargs)
+
+def DictWhenKeyExists(choices):
+    return Dict(when_key_exists=choices)
```
Comment (chrisgemignani) [inline sureberus/schema.py]: no kwargs?

File: test_sure.py [hunk 2]
```diff
@@ -395,3 +397,117 @@ def test_coerce_raises():
...
+        normalize_schema(choice_existence_schema, v)
+
+def test_when_key_exists_error_multiple_keys_exist():
+    v = {'image': 'foo', 'width': 3, 'pattern': {}, 'color': 'red'}
+    with pytest.raises(E.DisallowedField) as ei:
+        normalize_schema(choice_existence_schema, v)
+    assert {ei.value.field, ei.value.excluded} == {'image', 'pattern'}
+
+def test_when_key_exists_NO_keys_exist():
+    v = {'width': 30}
```
Comment (chrisgemignani) [inline test_sure.py]: It seems like we would want to support common schema for when_key_exists. For instance I'm able to do this

```choice_schema = S.DictWhenKeyIs(
    'type',
    {
        'foo': {'schema': {'foo_sibling': S.String()}},
        'bar': {'schema': {'bar_sibling': S.Integer()}},
    },
    schema={'sibling': S.String(required=False)}
)
``` 

but can't do this for WhenKeyExists due to lack of **kwargs

File: sureberus/schema.py [hunk 3]
```diff
@@ -11,11 +11,19 @@ def mk(d, kw, **morekw):
     schema.update(morekw)
     return schema
 
-def Dict(required=True, anyof=None, schema={}, **kwargs):
+def Dict(required=True, anyof=None, schema=None, **kwargs):
+    if schema is None:
```
Comment (chrisgemignani) [inline sureberus/schema.py]: :+1:

File: sureberus/schema.py (hunk 1 above)
Comment (radix) [reply sureberus/schema.py]: woops, thanks!

File: test_sure.py (hunk 2 above)
Comment (radix) [reply test_sure.py]: I think we're on the same page, but just to make sure, this *is* possible as shown by `test_when_key_exists_common_schema`, it's just the DictWhenKeyExists helper function doesn't pass through kwargs. I'll fix that, thank you for pointing it out!

**Comment by radix** [comment]
I've added the `**kwargs` passthrough to DictWhenKeyExists`, thanks!

File: test_sure.py (hunk 2 above)
Comment (chrisgemignani) [reply test_sure.py]: Yup, I found that example after making this comment :-P

**Comment by chrisgemignani** [comment]
:+1: Looked good

**Pull request merged**